dll/win32/msafd/misc/sndrcv.c

```diff
@@ -292,12 +298,12 @@ WSPRecv(SOCKET Handle,
                                    0);
 
     /* Wait for completion of not overlapped */
-    if (Status == STATUS_PENDING && lpOverlapped == NULL)
+    if (Status == STATUS_PENDING)
     {
         /* It's up to the protocol to time out recv.  We must wait
          * until the protocol decides it's had enough.
          */
-        WaitForSingleObject(SockEvent, INFINITE);
+        WaitForSingleObject(SockEvent, (lpOverlapped || Socket->SharedData->NonBlocking) ? 0 : INFINITE);
```

Comment:
With this change now the comment above feels contradictory to what the inner logic does right now as now the function returns immediately because it didn't wait due to the fact your forced the socket to be in a non blocking state.

Reply:
In any "non-blocking" case, you shouldn't go into this code path at all, because
* the wait with zero timeout does nothing
* reading `IOSB->Status` while the operation is still in progress means you have no guarantees about what value you'll get.
So just keep the `&& lpOverlapped == NULL` condition above, and add your `NonBlocking` condition in addition to that.

Reply:
Thanks @ThFabba for the explanation. So, as I understood, I should just add `&& !Socket->SharedData->NonBlocking` condition to the existing check. But the problem is, in this case, that code path will be called anyway, because `NonBlocking` is actually `FALSE` when `WSPRecv` function is called. So the condition still always succeeds.
I investigated the problem a bit, and now see that Firefox first creates the socket via calling `WSPSocket` function, and then passes its handle to `WSPRecv`. We're always creating blocking sockets by default, since `NonBlocking` is always set to `FALSE` in the first function: https://github.com/reactos/reactos/blob/master/dll/win32/msafd/misc/dllmain.c#L214. After changing it to `TRUE` manually, everything seems to work properly again.
But I'm not sure whether this is correct and want to inspect this member in Windows via WinDBG somehow, to check it exactly. But I don't know how. :frowning_face:

Reply:
Yes, the function is probably correct as-is, and your added code is just a workaround for some other limitation. My suggestion was mostly to make your workaround more concise and less questionable in terms of coding practices; it wouldn't change anything about it being a workaround.